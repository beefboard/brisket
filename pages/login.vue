<template>
  <div class="container">
    <form
      method="post"
      class="login"
      @submit.prevent="login">
      <input
        v-model="username"
        type="text"
        placeholder="Username">
      <input
        v-model="password"
        type="password"
        placeholder="Password">
      <button :disabled="loading">Login</button>
    </form>
  </div>
</template>

<script>
export default {
  data() {
    return {
      username: null,
      password: null,
      loading: false
    }
  },
  async asyncData({ store, redirect }) {
    if (store.state.auth) {
      redirect('/')
    }
  },
  methods: {
    async login() {
      this.loading = true
      try {
        await this.$store.dispatch('login', {
          username: this.username,
          password: this.password
        })
        await this.$router.push('/')
      } catch (e) {
        this.loading = false
        console.log(e)
        console.log('login failed')
      }
    }
  }
}
</script>

<style>
.container {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.login {
  display: flex;
  flex-direction: column;
}
</style>
