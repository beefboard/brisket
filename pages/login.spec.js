import { shallowMount } from '@vue/test-utils'
import login from './login'
import flushPromises from 'flush-promises'

describe('login', () => {
  test('mounts properly', () => {
    const wrapper = shallowMount(login)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  test('renders properly', () => {
    const wrapper = shallowMount(login)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should contain a login button', () => {
    const wrapper = shallowMount(login)
    expect(wrapper.find('button')).toBeTruthy()
  })

  test('login button should be disabled on load', () => {
    const wrapper = shallowMount(login)
    expect(wrapper.find('button').attributes('disabled')).toBe('disabled')
  })

  test('error message should not be visible on load', () => {
    const wrapper = shallowMount(login)
    expect(wrapper.find('.error').text()).toBe('')
  })

  test('entering username and password enables login button', () => {
    const wrapper = shallowMount(login)
    const usernameInput = wrapper.find('[placeholder="Username"]')
    const passwordInput = wrapper.find('[placeholder="Password"]')
    usernameInput.setValue('test')
    passwordInput.setValue('test')

    expect(wrapper.find('button').attributes('disabled')).toBe(undefined)
  })

  test('submitting form without credentials does nothing', async () => {
    const wrapper = shallowMount(login, {
      mocks: {
        $store: {
          dispatch: jest.fn()
        },
        $router: {
          push: jest.fn()
        }
      }
    })

    wrapper.find('form').trigger('submit')
    await flushPromises()

    expect(wrapper.vm.$store.dispatch).not.toHaveBeenCalled()
  })

  test('submitting form logs in with credentials and navigates to /', async () => {
    const wrapper = shallowMount(login, {
      mocks: {
        $store: {
          dispatch: jest.fn()
        },
        $router: {
          push: jest.fn()
        }
      }
    })

    const usernameInput = wrapper.find('[placeholder="Username"]')
    const passwordInput = wrapper.find('[placeholder="Password"]')
    usernameInput.setValue('test')
    passwordInput.setValue('test')

    wrapper.find('form').trigger('submit')
    await flushPromises()

    expect(wrapper.vm.$store.dispatch).toHaveBeenCalledWith('login', {
      username: 'test',
      password: 'test'
    })
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith('/')
  })

  it('should display error messages on failed login', async () => {
    const wrapper = shallowMount(login, {
      mocks: {
        $store: {
          dispatch: jest.fn()
        },
        $router: {
          push: jest.fn()
        }
      }
    })

    let mockError = {}
    wrapper.vm.$store.dispatch.mockImplementation(async () => {
      throw mockError
    })

    const usernameInput = wrapper.find('[placeholder="Username"]')
    const passwordInput = wrapper.find('[placeholder="Password"]')

    // Test unauthorised
    mockError = {
      response: {
        status: 401
      }
    }
    usernameInput.setValue('test')
    passwordInput.setValue('test')
    wrapper.find('form').trigger('submit')

    await flushPromises()

    expect(wrapper.find('.error').html()).toContain(
      'Invalid username or password'
    )
    // The password should be set to none
    expect(wrapper.vm.password).toBe('')

    mockError = {
      response: {
        status: 500
      }
    }
    usernameInput.setValue('test')
    passwordInput.setValue('test')
    wrapper.find('form').trigger('submit')

    await flushPromises()

    expect(wrapper.find('.error').html()).toContain('Server error')

    mockError = new Error('Something else')
    usernameInput.setValue('test')
    passwordInput.setValue('test')
    wrapper.find('form').trigger('submit')

    await flushPromises()

    expect(wrapper.find('.error').html()).toContain('Unknown error')

    mockError = mockError = {
      response: {
        status: -1
      }
    }
    usernameInput.setValue('test')
    passwordInput.setValue('test')
    wrapper.find('form').trigger('submit')

    await flushPromises()

    expect(wrapper.find('.error').html()).toContain('Unknown error')
  })

  it('should set title to "Login - Beefboard"', () => {
    const wrapper = shallowMount(login)
    const headData = wrapper.vm.$options.head()

    expect(headData.title).toBe('Login - Beefboard')
  })

  it('should redirect to home if already logged in', () => {
    const wrapper = shallowMount(login)

    const redirect = jest.fn()
    const mockStore = {
      state: {
        auth: {
          username: 'sdfsdf'
        }
      }
    }

    wrapper.vm.$options.asyncData({ redirect, store: mockStore })
    expect(redirect).toHaveBeenCalledWith('/')
  })

  it('should not redirect if not logged in', () => {
    const wrapper = shallowMount(login)

    const redirect = jest.fn()
    const mockStore = {
      state: {
        auth: null
      }
    }

    wrapper.vm.$options.asyncData({ redirect, store: mockStore })
    expect(redirect).not.toHaveBeenCalled()
  })
})
