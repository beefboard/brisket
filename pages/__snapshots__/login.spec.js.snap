// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login renders properly 1`] = `
<div class="container">
  <form method="post" class="login">
    <div><input type="text" placeholder="Username"> <input type="password" placeholder="Password">
      <!---->
    </div> <button disabled="disabled" type="submit">
      Login
    </button>
  </form>
</div>
`;
